### ISSUE TITLE
Guidance on How to Tune BatchedDataLoader

### ISSUE THREAD

**Issue by andrewredd** [uber/petastorm#570]
Hi All,

I tried reading the docs, issues, and pull requests and haven't felt like I have a good handle on how to optimize a dataloader in pytorch. I'm using Pytorch 1.5, Cloudera HDFS, Nvidia Tesla V100, have 80 cores and 700 gb of memory. 

I'm trying to read a smallish table of 36M rows with ~ 2000 values per row (they are grouped into five array columns). I've made sure the parquet block size is ~256MB

This is what I have for code:

`    dl = BatchedDataLoader(make_batch_reader(file_name, 
                                             workers_count=10, 
                                             schema_fields=col_names, 
                                             hdfs_driver='libhdfs',
                                             reader_pool_type = 'process'), 
                           batch_size=batch_size, 
                           shuffling_queue_capacity=shuffling_queue_capacity)`

Problem:
I'm trying to figure out how to keep a single GPU running at full memory capacity from the cluster. The setup above is majorly IO bound. In very basic tuning that it seems like I can get through an epoch faster using a large batch size (500K) and a shuffling_queue_capacity of 10x that. Even with the large batch size I only use up half of the GPU memory 17GB/32GB. This setup seems to process a batch every ~20secs, but then the GPU sits idle for ~30secs as it waits for another batch (I wish I knew how to benchmark all this better) I've experimented with larger and smaller shuffle queues but it seems to burn through the queue super fast and then be bound by IO out of the cluster.

I've tried a much smaller batch size (64) and petastorm cranks through the batches constantly but the GPU memory is minimally used.

Questions: 
1) Are large batches the best way to get high throughout out of the cluster?
2) Do I need to add more workers? I've increased the amount to 50 and haven't seen a improvement.
3) I read that the 'process' pool type is superior because I don't use any image c++ -like processing. Is this the right approach?
4) When I added the transform_fn to the batched data loader I'd almost immediately overflowed my gpu. Should I be making my batches smaller if I take this approach?

Would love to contribute a tutorial when we figure this out if that would be helpful.

The next step would be to put this on a much larger multi GPU machine. I need to figure out these throughput questions first. 

Thanks for the great work!

**Comment by andrewredd** [comment]
Seems similar to this issue https://github.com/uber/petastorm/issues/493

@selitvin do you have any guidance in terms of diagnosing the questions above?

**Comment by selitvin** [comment]
Sorry for the late response. I was away last week and was not monitoring my emails.

Thank you for the detailed description of your case and appreciate you going through existing issues and the documentation!

I'd ask the following questions:
- For your NN training code (forward/backward prop.) , what is the maximal row-per-sec consumption rate (let's call it r_nn)? What is the batchsize that you need to achieve that? One way to measure is to just cache the first batch and feed it repeatedly into your training code while measuring the throughput. You should get close to 100% GPU utilization in this experiment, I assume.
- What is the row-per-sec that you get from running the following tight-loop (even without rebatching):
```python
reader = make_batch_reader(...)
while True:
   batch = next(reader)  
   # measure row-per-sec  : let's call it r_io
   # Play with reader_pool_type process/thread; play with workers_count to see which combination gives you the highest r_io.
```

Do we have r_io > r_nn? 
- If r_io > r_nn, then we do have hope of setting up the system to saturate the GPU. As the next step I'd see how much rate do we drop by putting a `BatchedDataLoader` on top of the `make_batch_reader` and running it in the same tight loop. Are we loosing the throughput there?
- If r_io < r_nn, we'll need to profile what's going on in the IO land and how can it be optimized. To do so, I'd change `reader_pool_type='dummy'` (makes the worker run on the main thread), and run under Python profiler. See what are the hotspots and then think what can be done about it.

Hope this helps...

**Comment by andrewredd** [comment]
Thank you for the guidance @selitvin! Update on r_io

I've run the following code to benchmark r_io:

```
from itertools import product

pools = ['thread', 'process']
workers = [i for i in range(20, 81, 20)]
for i in product(pools, workers):
    print(i)

import time

from petastorm.reader import make_batch_reader
from petastorm.pytorch import BatchedDataLoader

for pool_type, workers_count in product(pools, workers):
    with make_batch_reader(file_name, 
                               workers_count=workers_count, 
                               schema_fields=schema_cols, 
                               hdfs_driver='libhdfs',
                               reader_pool_type = pool_type) as reader:
        
        r_io_sum = 0
        samples = 5
        for sample_idx in range(samples):
            num_rows = 0
            post_first_batch = False
            print(f'SAMPLE: {sample_idx} -- workers_count: {workers_count} and pool_type: {pool_type}')
            batch_idx = 0
            while True:
                batch = next(reader)
                
                # exclude the first batch from time computation in case
                # reader lags on first batch of reader
                if not post_first_batch:
                    post_first_batch = True
                    start = time.time()

                # print batches to make sure the process isnt hanging
                if batch_idx % 10 == 0:
                    print(batch_idx)

                # sum total number of rows to calculate throughput
                num_rows += batch[0].shape[0]
                
                if batch_idx > 50:
                    break
                    
                batch_idx += 1
                
            end = time.time()
            r_io = num_rows/(end - start)
            print(f'\t r_io: {r_io}')
            r_io_sum += r_io
        print(f'FINISHED: average r_io was {r_io_sum / samples}')
```

Which yielded the following output:

```
SAMPLE: 0 -- workers_count: 20 and pool_type: thread
   r_io: 44475.58341663292
SAMPLE: 1 -- workers_count: 20 and pool_type: thread
   r_io: 36072.09342879046
SAMPLE: 2 -- workers_count: 20 and pool_type: thread
   r_io: 41814.77067062044
SAMPLE: 3 -- workers_count: 20 and pool_type: thread
   r_io: 38848.62691198415
SAMPLE: 4 -- workers_count: 20 and pool_type: thread
   r_io: 31742.510717996367
FINISHED: average r_io was 38590.71702920487

SAMPLE: 0 -- workers_count: 40 and pool_type: thread
   r_io: 25514.446330343613
SAMPLE: 1 -- workers_count: 40 and pool_type: thread
   r_io: 20856.462452831598
SAMPLE: 2 -- workers_count: 40 and pool_type: thread
   r_io: 20377.239905442213
SAMPLE: 3 -- workers_count: 40 and pool_type: thread
   r_io: 21514.301401236044
SAMPLE: 4 -- workers_count: 40 and pool_type: thread
   r_io: 24267.352463445484
FINISHED: average r_io was 22505.96051065979

SAMPLE: 0 -- workers_count: 60 and pool_type: thread
   r_io: 5266.053699963165
SAMPLE: 1 -- workers_count: 60 and pool_type: thread
   r_io: 4043.862873196959
SAMPLE: 2 -- workers_count: 60 and pool_type: thread
   r_io: 5089.411847664342
SAMPLE: 3 -- workers_count: 60 and pool_type: thread
   r_io: 3746.3483067578336
SAMPLE: 4 -- workers_count: 60 and pool_type: thread
   r_io: 6092.0097035778335
FINISHED: average r_io was 4847.537286232027

SAMPLE: 0 -- workers_count: 80 and pool_type: thread
   r_io: 8447.037476335698
SAMPLE: 1 -- workers_count: 80 and pool_type: thread
   r_io: 5746.057307212415
SAMPLE: 2 -- workers_count: 80 and pool_type: thread
   r_io: 5697.119446744477
SAMPLE: 3 -- workers_count: 80 and pool_type: thread
   r_io: 4710.551820207388
SAMPLE: 4 -- workers_count: 80 and pool_type: thread
   r_io: 6087.477173457679
FINISHED: average r_io was 6137.648644791531

SAMPLE: 0 -- workers_count: 20 and pool_type: process
   r_io: 24855.088784545955
SAMPLE: 1 -- workers_count: 20 and pool_type: process
   r_io: 25643.679155180373
SAMPLE: 2 -- workers_count: 20 and pool_type: process
   r_io: 29341.33531734155
SAMPLE: 3 -- workers_count: 20 and pool_type: process
   r_io: 22948.268626977842
SAMPLE: 4 -- workers_count: 20 and pool_type: process
   r_io: 26578.438620526387
FINISHED: average r_io was 25873.36210091442

SAMPLE: 0 -- workers_count: 40 and pool_type: process
   r_io: 27265.894980715046
SAMPLE: 1 -- workers_count: 40 and pool_type: process
   r_io: 22653.8323385756
SAMPLE: 2 -- workers_count: 40 and pool_type: process
   r_io: 24016.25060442569
SAMPLE: 3 -- workers_count: 40 and pool_type: process
   r_io: 29931.950225921675
SAMPLE: 4 -- workers_count: 40 and pool_type: process
   r_io: 25301.035463043765
FINISHED: average r_io was 25833.792722536353

SAMPLE: 0 -- workers_count: 60 and pool_type: process
   r_io: 23639.409738587972
SAMPLE: 1 -- workers_count: 60 and pool_type: process
   r_io: 25768.323935785647
SAMPLE: 2 -- workers_count: 60 and pool_type: process
   r_io: 27432.125953724742
SAMPLE: 3 -- workers_count: 60 and pool_type: process
   r_io: 24085.147547309563
SAMPLE: 4 -- workers_count: 60 and pool_type: process
   r_io: 22382.420385521214
FINISHED: average r_io was 24661.485512185827

SAMPLE: 0 -- workers_count: 80 and pool_type: process
   r_io: 21542.793806437763
SAMPLE: 1 -- workers_count: 80 and pool_type: process
   r_io: 22618.27064951612
SAMPLE: 2 -- workers_count: 80 and pool_type: process
   r_io: 28569.499171402244
SAMPLE: 3 -- workers_count: 80 and pool_type: process
   r_io: 20477.581421107334
SAMPLE: 4 -- workers_count: 80 and pool_type: process
   r_io: 28457.16819380502
FINISHED: average r_io was 24333.062648453695
```

The code takes five samples of each run because 50 batches returned a different throughput (depending I imagine on the availability of the cluster). Interestingly the grid search over performance indicated that the thread approach with 20 workers outperfromed a process approach with more workers or more workers on the thread approach. Does this indicate a bug in the code? I'm showing consistent results across runs that reach nearly double the throughput that I currently have with 80 workers and process pooling. 

Going to move on to r_nn next.

**Comment by selitvin** [comment]
As for the benchmarking code, I'd suggest having several "warmup" cycles - there are some queues that are being filled up in the beginning, and you might want to measure the steady-state of the system - after all queues are full. 

You might want to setup `num_epochs=20` (or whatever number make sense) and not recreated reader in-between.

For this benchmark it make sense that thread-pool is faster as the inter-process-communication with a process pool is implemented in pyzmq. Although fast, it can not beat in-process thread-pool. It's great that thread-pool is faster because it means we don't do much of Python GIL intensive operations.

Things might start looking worse (hopefully not by much) when we will go through `BatchedDataLoader`. That's another interesting data point.

**Comment by andrewredd** [comment]
Is there a way to alter the process pool or the number of workers without
recreating the reader? I think the reader creation slows down in later
iterations because the work pool isn’t getting cleared because I’m in
Jupyter.

**Comment by selitvin** [comment]
Oops, I misread your code. A reader recreation is critical as you change the pool-type/workers count. I think your code is correct and the benchmark properly estimate the r_io.
